stripped/colabs/Birch_stripped.py

```diff
@@ -0,0 +1,732 @@
+  - czy_step_to_pustynia - wartość logiczna, gdy TRUE to traktujemy step jako pustynię.
+  """
+  fig, axes = plt.subplots(ncols=3, figsize=(18, 6))
+
+  scaled_df_train = scaled_df.loc[:,'Rainf':'SoilM_100_200cm']
+
+  for perplexity, ax in zip(list_of_perplexities, axes.flat):
+    tsne = TSNE(n_components = 2, perplexity = perplexity, n_jobs=-1, random_state=2024)
+    tsne_results = tsne.fit_transform(scaled_df_train)
+
+    df = pd.DataFrame()
+    df['dim 1'] = tsne_results[:,0]
+    df['dim 2'] = tsne_results[:,1]
+    if czy_step_to_pustynia:
+      df['Cluster'] = scaled_df["pustynia_i_step"]
+    else:
+      df['Cluster'] = scaled_df["pustynia"]
+    sns.scatterplot(x="dim 1", y="dim 2", hue="Cluster", palette = ["darkgreen", "orange", "blue"], data=df, legend="full", ax=ax).set(title=f'Perplexity: {perplexity}, Divergence: {round(tsne.kl_divergence_,3)}')
+
+
+# In[ ]:
+
+
+def plot_3d_tsne(perplexity: float, scaled_df: pd.DataFrame, original_df: pd.DataFrame, anotated_df: pd.DataFrame, czy_step_to_pustynia: bool=False) -> None:
+  """
+  Funkcja tworzy wykres 3d dla wybranej wartości perplexity.
+  Parametry:
+  - list_of_perplexities - lista wartości perplexity dla których chcemy stworzyć wykresy,
+  - scaled_df - ramka danych z już przeskalowanymi wartościami na podstawie której wykonujemy algorytm t-SNE,
+  - df_with_target - ramka danych w której znajdują się kolumny "pustynia", "step", "lon" i "lat",
+  - czy_step_to_pustynia - wartość logiczna, gdy TRUE to traktujemy step jako pustynię.
+  """
+  tsne = TSNE(n_components = 3, perplexity = perplexity)
+  tsne_results = tsne.fit_transform(scaled_df)
+
+  df_merged_with_anotated = original_df.merge(anotated_df[['lon', 'lat', 'pustynia', 'step', 'pustynia_i_step']], left_on=['lon','lat'], right_on = ['lon','lat'], how='left')
+  df_merged_with_anotated.fillna(2, inplace=True)
+
+  df = pd.DataFrame()
+  df['dim 1'] = tsne_results[:,0]
+  df['dim 2'] = tsne_results[:,1]
+  df['dim 3'] = tsne_results[:,2]
+  if czy_step_to_pustynia:
+    df['Cluster'] = df_merged_with_anotated["pustynia_i_step"]
+  else:
+    df['Cluster'] = df_merged_with_anotated["pustynia"]
+
+  fig = go.Figure(data=[go.Scatter3d(
+    x=df['dim 1'],
+    y=df['dim 2'],
+    z=df['dim 3'],
+    mode="markers",
+    marker=dict(color=df['Cluster'], colorscale=["darkgreen", "orange"]),
+    text = [f"Współrzędne: ({lat}, {lon})" for lat, lon in df_merged_with_anotated[["lat", "lon"]].values],
+    hoverinfo='text'
+  )])
+  fig.update_layout(scene = dict(
+                    xaxis_title='dim 1',
+                    yaxis_title='dim 2',
+                    zaxis_title='dim 3'))
+  fig.update_layout(scene=dict(xaxis_showspikes=False,
+                             yaxis_showspikes=False))
+  fig.show()
+
+
+# In[ ]:
+
+
+def do_kmeans_with_tsne_return_preds(list_of_scaled_df: List, n_components: int=2, perplexity:float=30, learning_rate: Union[float, str]="auto", n_clusters: int=2, random_state: int=123) -> List:
```

Comment:
Typing dla list_of_scaled_df to chyba List[pd.DataFrame] ? Btw. Może lepiej byłoby, żeby funkcja robiła to co robi dla pojedynczego dataframe a nie całej listy dataframe'ów? (i potem zastosowac ja dla poszczególnych df)